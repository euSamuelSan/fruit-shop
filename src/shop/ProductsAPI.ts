import apple from '../assets/images/apple.png'
import banana from '../assets/images/banana.png'
import mango from '../assets/images/mango.png'
import pear from '../assets/images/pear.png'
import pineapple from '../assets/images/pineapple.png'

export type Fruit = {
    id: string
    name: string
    price: number
    image: string
    measurementUnit: string
}

type GetAllResponse = {
    data: Fruit[]
}

type GetResponse = {
    data: Fruit
}

export const getAllFruits = async (): Promise<GetAllResponse | void> => {
    try {
        return Promise.resolve({ data: fruitsMock })
    } catch (err) {
        console.error(err)
    }
}

export const getFruitById = async (id: string): Promise<GetResponse | void> => {
    try {
        const item = fruitsMock.find(fruit => fruit.id === id)

        return item
            ? Promise.resolve({ data: item })
            : Promise.reject('ID informado não encontrado.')
    } catch (err) {
        console.error(err)
    }
}

export const fruitsMock: Fruit[] = [
    {
        id: '1',
        name: 'Maçã',
        price: 5.56,
        image: apple,
        measurementUnit: 'KG',
    },
    {
        id: '2',
        name: 'Pêra',
        price: 7.22,
        image: pear,
        measurementUnit: 'KG',
    },
    {
        id: '3',
        name: 'Banana',
        price: 2.5,
        image: banana,
        measurementUnit: 'KG',
    },
    {
        id: '4',
        name: 'Abacaxi',
        price: 6.65,
        image: pineapple,
        measurementUnit: 'Un',
    },
    {
        id: '5',
        name: 'Manga',
        price: 4.44,
        image: mango,
        measurementUnit: 'KG',
    },
]
