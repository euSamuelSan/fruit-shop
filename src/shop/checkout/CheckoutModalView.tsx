import React from 'react'
import styled from 'styled-components'
import { FiTrash } from 'react-icons/fi'
import { BsBagPlus, BsCreditCard2Back } from 'react-icons/bs'

import Colors from '../../commons/Colors'
import Modal from '../../commons/components/Modal'
import ActionButton from '../../commons/components/ActionButton'
import ItemOverviewComponent from './ItemOverviewComponent'
import { Fruit } from '../ProductsAPI'
import { standardizePrice } from '../../commons/StringUtils'

type ProductWithQuantity = Fruit & {
    quantity: number
}

export type CheckoutModalViewProps = {
    isModalVisible: boolean
    setIsModalVisible: React.Dispatch<React.SetStateAction<boolean>>
    items: ProductWithQuantity[]
    clearCart: () => void
    getTotalPrice: () => number
}
export default function CheckoutModalView({
    isModalVisible,
    setIsModalVisible,
    items,
    clearCart,
    getTotalPrice,
}: CheckoutModalViewProps): JSX.Element {
    const renderNoContent = (): JSX.Element => {
        return (
            <NoContentContainer>
                <NoContentIcon size={64} color={Colors.hospitalGreen} />
                <NoContentText>
                    Parece que você não possui produtos no carrinho. <br />
                    Volte para a loja para adicionar mais itens!
                </NoContentText>
            </NoContentContainer>
        )
    }

    const renderItemOverview = (
        item: ProductWithQuantity,
        index: number
    ): JSX.Element => {
        return (
            <ItemOverviewComponent product={item} key={item.id} index={index} />
        )
    }

    const renderItems = (): JSX.Element => {
        return (
            <Container>
                <ItemsGridHeader>
                    <ItemsGridHeaderText></ItemsGridHeaderText>
                    <ItemsGridHeaderText>Produto</ItemsGridHeaderText>
                    <ItemsGridHeaderText>Subtotal</ItemsGridHeaderText>
                    <ItemsGridHeaderText></ItemsGridHeaderText>
                </ItemsGridHeader>
                {items.map((item, index) => renderItemOverview(item, index))}
            </Container>
        )
    }

    return (
        <Modal
            title="Checkout"
            isVisible={isModalVisible}
            setIsVisible={setIsModalVisible}
        >
            <>
                {items.length ? renderItems() : renderNoContent()}
                {items.length ? (
                    <Footer>
                        <TotalPriceText>
                            Total: {standardizePrice(getTotalPrice())}
                        </TotalPriceText>
                        <ActionButton
                            text="Limpar Carrinho"
                            onClick={() => clearCart()}
                            color={Colors.strawberry}
                            icon={
                                <FiTrash size={16} color={Colors.strawberry} />
                            }
                        />
                        <ActionButton
                            text="Finalizar Pedido"
                            onClick={() => {
                                //TODO
                            }}
                            icon={
                                <BsCreditCard2Back
                                    size={16}
                                    color={Colors.white}
                                />
                            }
                            primary
                        />
                    </Footer>
                ) : null}
            </>
        </Modal>
    )
}

const NoContentContainer = styled.div`
    padding-top: 100px;
    display: flex;
    flex-direction: column;
    align-items: center;
`

const NoContentText = styled.p`
    font-size: 12;
    color: ${Colors.lightBlueGreyTwo};
    text-align: center;
`
const NoContentIcon = styled(BsBagPlus)``

const ItemsGridHeader = styled.div`
    display: grid;
    grid-template-columns: 0.4fr 0.7fr 1.9fr 1fr;
    padding: 0 10px;
    height: 40px;
`

const ItemsGridHeaderText = styled.p`
    color: ${Colors.black};
    font-weight: bold;
    margin: 0;
`

const Container = styled.div`
    padding: 10px 0;
    display: flex;
    flex-direction: column;
    height: max-content;
`

const Footer = styled.div`
    width: 480px;
    margin-top: auto;
    align-self: flex-end;
    flex-direction: flex-end;
    display: flex;
    gap: 5px;
    padding: 10px;
`

const TotalPriceText = styled.p`
    color: ${Colors.black};
    font-weight: 500;
    margin: 0 auto 0 0;
`
