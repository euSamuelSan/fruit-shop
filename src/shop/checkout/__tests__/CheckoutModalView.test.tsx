import { fireEvent, render, RenderResult, screen } from '@testing-library/react'

import CheckoutModalView, { CheckoutModalViewProps } from '../CheckoutModalView'
import { fruitsMock } from '../../ProductsAPI'
import { act } from 'react-dom/test-utils'

describe('CheckoutModalView', () => {
    const getProps = (): CheckoutModalViewProps => ({
        isModalVisible: true,
        setIsModalVisible: jest.fn(),
        items: [],
        clearCart: jest.fn(),
        getTotalPrice: jest.fn(() => 1),
    })

    const renderCheckoutModalView = (props = getProps()): RenderResult => {
        return render(<CheckoutModalView {...props} />)
    }

    it('Deve desenhar o componente corretamente', () => {
        renderCheckoutModalView()

        expect(document.body).toMatchSnapshot()
    })

    it('Deve desenhar o componente com produtos corretamente', () => {
        const props = getProps()
        props.items = [{ ...fruitsMock[0], quantity: 2 }]

        renderCheckoutModalView(props)

        expect(document.body).toMatchSnapshot()
    })

    it('Deve limpar o carrinho ao clicar no botão', () => {
        const props = getProps()
        props.items = [{ ...fruitsMock[0], quantity: 2 }]

        renderCheckoutModalView(props)
        const clearButton = screen.getByText('Limpar Carrinho')

        fireEvent.click(clearButton)
        expect(props.clearCart).toHaveBeenCalled()
    })

    it('Deve fechar a modal ao clicar no botão', () => {
        const props = getProps()
        props.items = [{ ...fruitsMock[0], quantity: 2 }]

        renderCheckoutModalView(props)
        const [closeButton] = screen.getAllByRole('button')

        fireEvent.click(closeButton)
        expect(props.setIsModalVisible).toHaveBeenCalledTimes(1)
        expect(props.setIsModalVisible).toHaveBeenCalledWith(false)
    })
})
