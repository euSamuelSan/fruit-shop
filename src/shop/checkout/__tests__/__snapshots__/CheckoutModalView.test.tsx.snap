// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutModalView Deve desenhar o componente com produtos corretamente 1`] = `
<body>
  <div>
    <div
      class="sc-bczRLJ iVAlPu"
    >
      <div
        class="sc-gsnTZi iIKTfW"
        height="400px"
        width="500px"
      >
        <div
          class="sc-dkzDqf gtawzm"
        >
          <p
            class="sc-hKMtZM kYtwUi"
          >
            Checkout
          </p>
          <button
            class="sc-eCYdqJ gHfvpn"
          >
            <svg
              fill="currentColor"
              height="20"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
              />
            </svg>
          </button>
        </div>
        <div
          class="sc-fnykZs ceFtpf"
        >
          <div
            class="sc-ksZaOG dNKrkI"
          >
            <p
              class="sc-hAZoDl icLmXY"
            />
            <p
              class="sc-hAZoDl icLmXY"
            >
              Produto
            </p>
            <p
              class="sc-hAZoDl icLmXY"
            >
              Subtotal
            </p>
            <p
              class="sc-hAZoDl icLmXY"
            />
          </div>
          <div
            class="sc-iBkjds hPDXXX"
          >
            <img
              alt="Maçã"
              class="sc-ftvSup fapAys"
              src="apple.png"
            />
            <p
              class="sc-papXJ eeksmv"
            >
              Maçã
            </p>
            <p
              class="sc-papXJ eeksmv"
            >
              <b>
                R$ 11,12
              </b>
               (
              R$ 5,56
               por 
              KG
              )
            </p>
            <div
              class="sc-kDDrLX jbqpdq"
            >
              <button
                class="sc-iqcoie rqbip"
              >
                <svg
                  color="#8e8e93"
                  fill="currentColor"
                  height="20"
                  stroke="currentColor"
                  stroke-width="0"
                  style="color: rgb(142, 142, 147);"
                  viewBox="0 0 1024 1024"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title>
                    Remover
                  </title>
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm192 472c0 4.4-3.6 8-8 8H328c-4.4 0-8-3.6-8-8v-48c0-4.4 3.6-8 8-8h368c4.4 0 8 3.6 8 8v48z"
                  />
                </svg>
              </button>
              <p
                class="sc-papXJ sc-jqUVSM eeksmv kispeS"
              >
                2
                 
                KG
              </p>
              <button
                class="sc-iqcoie rqbip"
              >
                <svg
                  color="#8e8e93"
                  fill="currentColor"
                  height="20"
                  stroke="currentColor"
                  stroke-width="0"
                  style="color: rgb(142, 142, 147);"
                  viewBox="0 0 1024 1024"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title>
                    Adicionar
                  </title>
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm192 472c0 4.4-3.6 8-8 8H544v152c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V544H328c-4.4 0-8-3.6-8-8v-48c0-4.4 3.6-8 8-8h152V328c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v152h152c4.4 0 8 3.6 8 8v48z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="sc-fEOsli eWTHOO"
        >
          <p
            class="sc-bjUoiL dCKJQI"
          >
            Total: 
            R$ 1,00
          </p>
          <button
            class="sc-jSMfEi dIpffO"
            color="#eb394b"
          >
            <svg
              color="#eb394b"
              fill="none"
              height="16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="color: rgb(235, 57, 75);"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="3 6 5 6 21 6"
              />
              <path
                d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
              />
            </svg>
            <p
              class="sc-gKXOVf xXklM"
            >
              Limpar Carrinho
            </p>
          </button>
          <button
            class="sc-jSMfEi gRQbRT"
            color="#57bd41"
          >
            <svg
              color="#ffffff"
              fill="currentColor"
              height="16"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(255, 255, 255);"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11 5.5a.5.5 0 0 1 .5-.5h2a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1-.5-.5v-1z"
              />
              <path
                d="M2 2a2 2 0 0 0-2 2v8a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V4a2 2 0 0 0-2-2H2zm13 2v5H1V4a1 1 0 0 1 1-1h12a1 1 0 0 1 1 1zm-1 9H2a1 1 0 0 1-1-1v-1h14v1a1 1 0 0 1-1 1z"
              />
            </svg>
            <p
              class="sc-gKXOVf xXklM"
            >
              Finalizar Pedido
            </p>
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`CheckoutModalView Deve desenhar o componente corretamente 1`] = `
<body>
  <div>
    <div
      class="sc-bczRLJ iVAlPu"
    >
      <div
        class="sc-gsnTZi iIKTfW"
        height="400px"
        width="500px"
      >
        <div
          class="sc-dkzDqf gtawzm"
        >
          <p
            class="sc-hKMtZM kYtwUi"
          >
            Checkout
          </p>
          <button
            class="sc-eCYdqJ gHfvpn"
          >
            <svg
              fill="currentColor"
              height="20"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
              />
            </svg>
          </button>
        </div>
        <div
          class="sc-crXcEl iCOoPH"
        >
          <svg
            class="sc-breuTD"
            color="#a3e9a4"
            fill="currentColor"
            height="64"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(163, 233, 164);"
            viewBox="0 0 16 16"
            width="64"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 7.5a.5.5 0 0 1 .5.5v1.5H10a.5.5 0 0 1 0 1H8.5V12a.5.5 0 0 1-1 0v-1.5H6a.5.5 0 0 1 0-1h1.5V8a.5.5 0 0 1 .5-.5z"
              fill-rule="evenodd"
            />
            <path
              d="M8 1a2.5 2.5 0 0 1 2.5 2.5V4h-5v-.5A2.5 2.5 0 0 1 8 1zm3.5 3v-.5a3.5 3.5 0 1 0-7 0V4H1v10a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V4h-3.5zM2 5h12v9a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V5z"
            />
          </svg>
          <p
            class="sc-evZas fRLgxg"
          >
            Parece que você não possui produtos no carrinho. 
            <br />
            Volte para a loja para adicionar mais itens!
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
